2022Swerve/src/main/java/frc/robot/hardware/AbsoluteEncoders.java

```diff
@@ -0,0 +1,69 @@
+package frc.robot.hardware;
+
+import java.util.function.Supplier;
+
+import edu.wpi.first.wpilibj.AnalogInput;
+import edu.wpi.first.wpilibj.RobotController;
+
+public enum AbsoluteEncoders {
+    //Swerve Modules
+    FrontLeftModule(9),
+    FrontRightModule(10),
+    BackLeftModule(11),
+    BackRightModule(12);
+    
+    private int ID;
+    private boolean reversed;
+    private double offset; //Offset in radians
+    private AnalogInput encoder;
+    private Supplier<Double> supplier;
+
+    AbsoluteEncoders(int ID, boolean reversed, double offset){
+        this.ID = ID;
+        this.reversed = reversed;
+        this.offset = offset;
+        encoder = new AnalogInput(this.ID);
+        initializeSupplier();
+    }
+
+    AbsoluteEncoders(int ID, boolean reversed){
+        this.ID = ID;
+        this.reversed = reversed;
+        this.offset = 0;
+        encoder = new AnalogInput(this.ID);
+        initializeSupplier();
+    }
+
+    AbsoluteEncoders(int ID){
+        this.ID = ID;
+        this.reversed = false;
+        this.offset = 0;
+        encoder = new AnalogInput(this.ID);
+        initializeSupplier();
+    }
+
+    private void initializeSupplier() {
+        supplier = new Supplier<Double>() {
+            public Double get() {
+                return calculateAngleRadians();
+            }
+        }; 
+    }
+
+    private double calculateAngleRadians(){ 
+        //Calculate the value of the absolute encoder in radians
+        double angle = encoder.getVoltage() / RobotController.getVoltage5V();
+        angle *= 2.0 * Math.PI;
+        angle -= offset;
+        return angle * (reversed ? -1.0 : 1.0);
+    }
+
+    public AnalogInput getEncoder() {
+        return encoder;
+    }
+
+    public Supplier<Double> getSupplier() { 
+        //Use this supplier to get the encoder value in radians
+        return supplier;
```

Comment:
Shouldn't this be:

return () -> calculateAngleRadians();

And initializeSupplier wouldn't exist?

Also, is there a better name for getSupplier?  It isn't descriptive unless you are in this part of the code

Reply:
`return () -> calculateAngleRadians();` What does a lambda do in this context?

I don't see a problem with getSupplier, it states clearly what the method does; returns the encoder's associated supplier. If I were getting the _value_ of the supplier, I would name it something more fitting (getAngleRadians or similar), but the method simply returns a supplier that can be later referenced.

Reply:
The lambda creates a function that is returned immediately, instead of creating a function upon initialization.

If you were reviewing this code, and hadn't written it.  Which would be clearer:  getAngleRadians or getSupplier?

Reply:
> The lambda creates a function that is returned immediately, instead of creating a function upon initialization.
So, am I understanding this correctly? Could the function returned here be instantiated as a variable in another class, so that whenever the variable is called it returns the current result of the function?

> If you were reviewing this code, and hadn't written it. Which would be clearer: getAngleRadians or
> getSupplier?
Depends on whether the method gets the angle in radians, or gets a supplier. The current method does not get anything _from_ the supplier, it simply gets the supplier itself.

Reply:
> So, am I understanding this correctly? Could the function returned here be instantiated as a variable in another class, so that whenever the variable is called it returns the current result of the function?


As I understand it, yes.